android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/KeyboardController.java

```diff
@@ -0,0 +1,196 @@
+      KMManager.KMFilename_KeyboardsList);
+    File keyboards_json = new File(context.getDir("userdata", Context.MODE_PRIVATE),
+      KMFilename_Installed_KeyboardsList);
+    if (list == null) {
+      list = new ArrayList<Keyboard>();
+      if (keyboards_dat.exists() && !keyboards_json.exists()) {
+        try {
+          // Migrate installed_keyboards.dat to installed_keyboards.json
+          ObjectInputStream inputStream = new ObjectInputStream(new FileInputStream(keyboards_dat));
+          ArrayList<HashMap<String, String>> dat_list = (ArrayList<HashMap<String, String>>) inputStream.readObject();
+          inputStream.close();
+
+          for(HashMap<String, String> kbdMap : dat_list) {
+            boolean isNewKeyboard = kbdMap.containsKey(KeyboardPickerActivity.KMKEY_INTERNAL_NEW_KEYBOARD) &&
+              kbdMap.get(KeyboardPickerActivity.KMKEY_INTERNAL_NEW_KEYBOARD).equals(KeyboardPickerActivity.KMKEY_INTERNAL_NEW_KEYBOARD);
+
+            Keyboard k = new Keyboard(
+              kbdMap.get(KMManager.KMKey_PackageID),
+              kbdMap.get(KMManager.KMKey_KeyboardID),
+              kbdMap.get(KMManager.KMKey_KeyboardName),
+              kbdMap.get(KMManager.KMKey_LanguageID),
+              kbdMap.get(KMManager.KMKey_LanguageName),
+              MapCompat.getOrDefault(kbdMap, KMManager.KMKey_Version, "1.0"),
+              MapCompat.getOrDefault(kbdMap, KMManager.KMKey_CustomHelpLink, ""),
+              isNewKeyboard,
+              MapCompat.getOrDefault(kbdMap, KMManager.KMKey_Font, null),
+              MapCompat.getOrDefault(kbdMap, KMManager.KMKey_OskFont, null)
+            );
+            list.add(k);
+          }
+
+        } catch (Exception e) {
+          Log.e(TAG, "Exception migrating installed_keyboards.dat");
+          list.add(Keyboard.DEFAULT_KEYBOARD);
+        }
+      } else if (keyboards_json.exists()) {
+        try {
+          // Get installed keyboards from installed_keyboards.json
+          JSONParser jsonParser = new JSONParser();
+          JSONArray json_list = jsonParser.getJSONObjectFromFile(keyboards_json, JSONArray.class);
+          if (json_list != null) {
+            // Can't foreach JSONArray
+            for (int i=0; i<json_list.length(); i++) {
+              JSONObject o = json_list.getJSONObject(i);
+              if (o != null) {
+                Keyboard k = new Keyboard(o);
+                list.add(k);
+              }
+            }
+          }
+        } catch (Exception e) {
+          Log.e(TAG, "Exception reading installed_keyboards.json");
+          list.add(Keyboard.DEFAULT_KEYBOARD);
+        }
+      } else {
+        // No installed keyboards lists so assume default
+        // TODO: What about 3rd-party apps w/o sil_euro_latin?
+        list.add(Keyboard.DEFAULT_KEYBOARD);
+      }
+
+      // We'd prefer not to overwrite a file if it exists
+      if (!keyboards_json.exists()) {
+        save(context);
```

Comment:
Can we delete the old keyboards_dat file once we've migrated?

Reply:
I'll mark it a TODO when the feature is stabilized